### ISSUE TITLE
Bug: API failed to parse valid OpenAPI/Swagger doc from a URL

### ISSUE THREAD

**Issue by globeone** [zaproxy/zaproxy#6091]
**Describe the bug**
One of the killer features of ZAP that no other scanner does is to scan all the calls of an API.
Unfortunately it seems to have a hard time reading valid OpenAPI/Swagger docs from a URL and the error message is not descriptive enough to give a clue as to what is going wrong. (although looking at the log the issue is deeper than what a regular user could solve)

**To Reproduce**
1. Go to Import... 
2. Click on Import an OpenAPI definition from a URL
3. Fill In the OpenAPI URL and the Target URL and press ok

**Expected behavior**
For ZAP to parse the OpenAPI definition and then whack the living Betelgeuse 🌟 out of the API until every API call had been tested and flagged for vulnerabilities.

**Screenshots**
![2020-07-22URLRESTImport](https://user-images.githubusercontent.com/10052587/88195402-4bce3580-cc40-11ea-807f-0b672731d9b6.png)

![2020-07-22_FailedToParseSwagger](https://user-images.githubusercontent.com/10052587/88193212-b03bc580-cc3d-11ea-92c7-5dc92da28ad0.png)

**Software versions**
 - ZAP: 2.9.0
 - Add-on: 
[2020-07-20ZAPInstalledAddOns.xlsx](https://github.com/zaproxy/zaproxy/files/4960771/2020-07-20ZAPInstalledAddOns.xlsx)

 - OS: Windows 10 1909 (18363.959)
 - Java: BellSoft Liberica 14
openjdk version "14.0.2" 2020-07-14
OpenJDK Runtime Environment (build 14.0.2+13)
OpenJDK 64-Bit Server VM (build 14.0.2+13, mixed mode, sharing)
 - Browser(s): 
Chrome Version 84.0.4147.89 (Official Build) (64-bit)
Firefox 78.0.2 (64-bits)

**Errors from the zap.log file**
I added the log as a text file as it's a bit too big for here.
[2020-07-22_SwaggerParserError.log.txt](https://github.com/zaproxy/zaproxy/files/4960734/2020-07-22_SwaggerParserError.log.txt)

**Additional context**
I've tried this on several valid OpenAPI swagger docs. The only time I've ever gotten it to work is by exporting the doc to JSON changing fields around until ZAP accepted the OpenAPI doc. 

**Would you like to help fix this issue?**
Yes, I have two valid professionally created OpenAPI docs on private URL's that I can test with. I believe that one of them is even publicly available on the Internet but I'll have to check. Once this feature works nearly flawless it'll be ZAP for the win.

**Comment by globeone** [comment]
This is one of the valid Swaggers I'm trying to import. Folded for brevity and security
![2020-07-22_ValidSwagger](https://user-images.githubusercontent.com/10052587/88195622-8768ff80-cc40-11ea-8db0-3a489eb4256f.png)

**Comment by thc202** [comment]
You need to import the actual OpenAPI definition (JSON or YAML), from the logs it's being imported HTML.

**Comment by globeone** [comment]
But there's an Option: import from URL this implies that the HTML Swagger Doc will be read.
![image](https://user-images.githubusercontent.com/10052587/88196292-4f15f100-cc41-11ea-9b57-2d77fc8fb755.png)

**Comment by globeone** [comment]
So maybe this is a feature request to Read the OpenAPI definition from HTML?

**Comment by globeone** [comment]
This is what we get from the vendor to scan with.
![2020-07-22SwaggerUI](https://user-images.githubusercontent.com/10052587/88196846-f2ff9c80-cc41-11ea-8633-e743038adaf1.png)

**Comment by thc202** [comment]
Importing form URL does not imply importing a random/unknown format (like HTML from Swagger UI), it's still to import an OpenAPI definition but from/in a HTTP server (vs importing from the local file system).

Don't you have the URL to the actual OpenAPI definition in the last example? (The URL used in the Swagger UI.)

**Comment by davewichers** [comment]
I don't think this is a reasonable request. You can't import an OpenAPI spec from HTML. HTML is the human readable API definition that is generated from a .yaml or .json file. yaml or json is the machine readable spec.

I think you should just close this issue.

**Comment by globeone** [comment]
Just speaking from real world web security scanning experience.
Then I must have misunderstood as I have never seen the JSON being available via a URL only the HTML. Every vendor offers the HTML rendering but no-one offers the JSON rendering. I've only had one internal developer who could deliver the JSON doc and they generated it mostly by hand and delivered it as a separate file to import to ZAP.
There must be a library as the HTML format is extremely standardized and tagged and should be reversible back to JSON for machine import. I'm going to dig a bit deeper. So if you humor me for a few days to do some research I'd like to keep this ticket open a bit longer.

Also, the dialog box and choices are then confusing as it matches exactly what one would expect to receive from a software developer or vendor in HTML. Where as the Import from File option would imply the JSON/YAML doc.

**Comment by globeone** [comment]
It might be easier than I previously thought.
https://stackoverflow.com/questions/41808417/how-to-generate-swagger-json
(still researching)

**Comment by kingthorin** [comment]
Most swagger UIs I've seen you can navigate to them as web pages but they still offer-up the actual JSON as basically the first link on the page.
Ex:
1. https://www.ebi.ac.uk/ena/submit/report/swagger-ui.html
  - Click "https://www.ebi.ac.uk/ena/submit/report/v2/api-docs"
2. https://dev.fitbit.com/build/reference/web-api/explore/#/Activity/getActivitiesByDate
  - Click "https://dev.fitbit.com/build/reference/web-api/explore/fitbit-web-api-swagger.json"
3. https://petstore.swagger.io/
  - Click "https://petstore.swagger.io/v2/swagger.json"

**Comment by globeone** [comment]
So thanks everyone for your help. 
OpenAPI makes it really difficult to get to the JSON which is not the fault of ZAP

ZAP could help though try to locate the openAPI JSON file
for example in the private API I'm testing it's a file called v1 (with no extension)
/REST/swagger/docs/v1

If there were some kind of automated Spider that could look for the JSON doc with the swagger header it would make it easier 
`{"swagger":"2.0","info": `

For example:
Please enter the Base URL: https://api.example.org/REST
Spider starts looks for a JSON file in the base URL starting with `{"swagger":"[0-9].0","info": `
Loads the API JSON definition.
And starts API related scans.

Or during a regular spider:
"We found an OpenAPI definition would you like to scan each of the API calls in this session?"

Or at least some better instruction in the dialog box 
instead of
"URL pointing to an OpenAPI defn" 
"URL pointing to an OpenAPI JSON definition"  this way there is no confusion as to which definition is expected.
![2020-07-22URLRESTImport](https://user-images.githubusercontent.com/10052587/88230971-a634b980-cc73-11ea-929d-30795ea24024.png)

Just some ideas to make ZAP easier to use.
May all be well and healthy.

**Comment by globeone** [comment]
For anyone else who may stumble upon this ticket in the future and need help finding the JSON.

- Go to the HTML version of your Swagger doc
- Open the browsers' developer console by typing
- CTRL+SHIFT+i
- Go to the Network tab
- then look for the document that is initiated by swagger-ui

![2020-07-22NetworkBrowserTraceToFindJSON](https://user-images.githubusercontent.com/10052587/88232094-91f1bc00-cc75-11ea-889d-01425a7785bf.png)

**Comment by globeone** [comment]
If I may be so bold. Would it also be possible to change the error handling when it doesn't find valid JSON or YAML. As it looks like the application recognizes that it doesn't have valid JSON, and knows why, but throws the end user in the deep end and says. Go find some log-file somewhere on your system that might help you. Go-Fetch. 🐕 
This even though ZAP knows exactly where it's writing its own log file so can easily open the log file for the end user.

The current error message can be improved to read:

**Failed to Parse OpenAPI definintion
Invalid JSON or YAML definition
Please enter a URL pointing to a valid OpenAPI JSON or YAML document. 
Press Open Log File to see the logs for more details.**
 
The Open Log File button could then either open the error log file in a Java viewer in a tab within ZAP, or use the system defined text editor depending on what's easier to program. 

![2020-07-23ErrorMessageImprovement](https://user-images.githubusercontent.com/10052587/88265806-6b13a400-ccce-11ea-8dba-c1af301c3a94.png)

**Comment by psiinon** [comment]
I'm all for more useful user error messages :)
The option to open the log file is interesting, and could be useful in lots of other situations. In dev mode we do expose error messages via the Output tab. Maybe we should do this in 'standard' mode too?

**Comment by thc202** [comment]
Tweaked title to match what's being asked/proposed.

> If there were some kind of automated Spider that could look for the JSON doc with the swagger header it would make it easier

The spider already does that, if it detects an OpenAPI definition it will import/spider it.

> Or at least some better instruction in the dialog box

I'd say that stating JSON is redundant (the "OpenAPI definition" already implies that is JSON/YAML, the formats specified in the OpenAPI spec) and misleading (the add-on can import YAML as well).

**Comment by globeone** [comment]
> I'd say that stating JSON is redundant (the "OpenAPI definition" already implies that is JSON/YAML, the formats specified in the OpenAPI spec) and misleading (the add-on can import YAML as well).

Redundancy is important here for clarity because OpenAPI definition always presents itself to the end user as HTML not as JSON nor YAML.
The JSON or YAML have to be hunted for as we have discussed earlier in the ticket.
